core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -383,11 +384,43 @@ class TransactionCoordinator(brokerId: Int,
 
   def partitionFor(transactionalId: String): Int = txnManager.partitionFor(transactionalId)
 
+  private def expireTransactions(): Unit = {
+
+    txnManager.transactionsToExpire().foreach{ idAndMetadata =>
+      idAndMetadata.metadata synchronized {
+        if (!txnManager.isCoordinatorLoadingInProgress(idAndMetadata.transactionalId)) {
+          idAndMetadata.metadata.producerEpoch = (idAndMetadata.metadata.producerEpoch + 1).toShort
+          idAndMetadata.metadata.prepareTransitionTo(Ongoing)
+          txnManager.appendTransactionToLog(idAndMetadata.transactionalId, idAndMetadata.metadata, (errors: Errors) => {
```

Comment:
Thinking about this a bit more, I can see the motivation for two phase approach: you will categorically fence off any existing producer while maintaining the `Ongoing` state. You can then safely begin the abort. 

However, even with the two phase approach, won't you still have a race condition where the existing producer could do a `PREPARE_COMMIT` which gets in before your epoch bump?